### PR TITLE
Fix overflow problem (#874)

### PR DESCRIPTION
**Pull request by yairzaslavsky**
An overflow occured while calculating the expoential backoff.

as MAX_RETRIES is 30, with baseDelay of 5000 you will get to an overflow
at the calculation baseDelay * (1 << retriesAttempted), a negative value will
be returned from Math.min , and passed as the delay value to the call
to Thread.sleep, causing an exception.

Comment for maintainers:
There was too much noise on my previous pull-request, I apologize for that, 
I submitted a pull request with your suggestions.

### CODE REVIEW

**Comment by kiiadi** [comment]
This has actually broken the tests - can you take a look?
https://travis-ci.org/aws/aws-sdk-java/jobs/166628623

Suspect it might be the change in logic from `random.nextInt(bound)` to `random.nextLong() % bound`

**Comment by yairzaslavsky** [comment]
@kiiadi  fixed, 

https://travis-ci.org/aws/aws-sdk-java/jobs/166906300  fails due to permGen issue.

File: aws-java-sdk-core/src/main/java/com/amazonaws/retry/PredefinedBackoffStrategies.java [hunk 1]
```diff
@@ -59,8 +59,7 @@ public FullJitterBackoffStrategy(final int baseDelay,
         public long delayBeforeNextRetry(AmazonWebServiceRequest originalRequest,
                                          AmazonClientException exception,
                                          int retriesAttempted) {
-            int ceil = (retriesAttempted > MAX_RETRIES) ? maxBackoffTime :
-                    Math.min(baseDelay * (1 << retriesAttempted), maxBackoffTime);
+            int ceil =  (int)calculateExponentialDelay(retriesAttempted, baseDelay, maxBackoffTime);
```
Comment (kiiadi) [inline aws-java-sdk-core/src/main/java/com/amazonaws/retry/PredefinedBackoffStrategies.java]: I'm not a huge fan of the casting between long and int - if we go over int's bound we're going to run into the same overflow problem when we do this cast.

File: aws-java-sdk-core/src/main/java/com/amazonaws/retry/PredefinedBackoffStrategies.java [hunk 2]
```diff
@@ -81,9 +80,8 @@ public EqualJitterBackoffStrategy(final int baseDelay,
         public long delayBeforeNextRetry(AmazonWebServiceRequest originalRequest,
                                         AmazonClientException exception,
                                         int retriesAttempted) {
-            int ceil = (retriesAttempted > MAX_RETRIES) ? maxBackoffTime
-                    : Math.min(maxBackoffTime, baseDelay * (1 << retriesAttempted));
-            return (ceil / 2) + random.nextInt((ceil / 2) + 1);
+            int ceil =  (int)calculateExponentialDelay(retriesAttempted, baseDelay, maxBackoffTime);
+            return (ceil / 2) + (long)random.nextInt((ceil / 2) + 1);
```
Comment (kiiadi) [inline aws-java-sdk-core/src/main/java/com/amazonaws/retry/PredefinedBackoffStrategies.java]: Do we need this cast to (long) ?

**Comment by kiiadi** [comment]
One other thing - I'd like to try to reproduce the overflow scenario you were seeing in a test case so that we're sure this has solved it and we don't regress in the future. Could you add a couple of test cases?

Thanks so much for the time spent on this - it'll be good to have this issue resolved.

**Comment by yairzaslavsky** [comment]
I was away for  some time and could not work on that, I decided to use a different approach. I will push a different approach + unit test.

**Comment by varunnvs92** [comment]
@yairzaslavsky 
We actually implemented a different approach internally as other customers were waiting for the fix. The fix will be included in tomorrow's release. Thank you for your effort and contribution.

**Comment by kiiadi** [comment]
It's worth pointing out that the approach we ended up going with was very similar to your most recent PR - so thanks for helping us find and fix this issue!

**Pull request closed without merging**